build.gradle

```diff
@@ -519,85 +521,101 @@ wrapper {
+                            "have the same name and group: ${coords}. " +
+                            "This doesn't currently work correctly in Gradle, see: " +
+                            "https://github.com/gradle/gradle/issues/847"
+            )
+        }
+    }
 }
 
 if (System.properties.get("build.compare") != null) {
-  apply plugin: 'compare-gradle-builds'
-  compareGradleBuilds {
-    ext.referenceProject = System.properties.get("build.compare")
-    doFirst {
-      if (file(referenceProject).exists() == false) {
-        throw new GradleException(
-                "Use git worktree to check out a version to compare against to ../elasticsearch_build_reference"
-        )
-      }
-    }
-    sourceBuild {
-      gradleVersion = gradle.getGradleVersion()
-      projectDir = referenceProject
-      tasks = ["clean", "assemble"]
-      arguments = ["-Dbuild.compare_friendly=true"]
-    }
-    targetBuild {
-      tasks = ["clean", "assemble"]
-      // use -Dorg.gradle.java.home= to alter jdk versions
-      arguments = ["-Dbuild.compare_friendly=true"]
-    }
-  }
+    apply plugin: 'compare-gradle-builds'
+    compareGradleBuilds {
+        ext.referenceProject = System.properties.get("build.compare")
+        doFirst {
+            if (file(referenceProject).exists() == false) {
+                throw new GradleException(
+                        "Use git worktree to check out a version to compare against to ../elasticsearch_build_reference"
+                )
+            }
+        }
+        sourceBuild {
+            gradleVersion = gradle.getGradleVersion()
+            projectDir = referenceProject
+            tasks = ["clean", "assemble"]
+            arguments = ["-Dbuild.compare_friendly=true"]
+        }
+        targetBuild {
+            tasks = ["clean", "assemble"]
+            // use -Dorg.gradle.java.home= to alter jdk versions
+            arguments = ["-Dbuild.compare_friendly=true"]
+        }
+    }
 }
 
 allprojects {
-  task resolveAllDependencies {
-      dependsOn tasks.matching { it.name == "pullFixture"}
-      doLast {
-        configurations.findAll { it.isCanBeResolved() }.each { it.resolve() }
-      }
-  }
+    task resolveAllDependencies {
+        dependsOn tasks.matching { it.name == "pullFixture" }
+        doLast {
+            configurations.findAll { it.isCanBeResolved() }.each { it.resolve() }
+        }
+    }
 }
 
 allprojects {
-  task checkPart1
-  task checkPart2 
-  tasks.matching { it.name == "check" }.all { check ->
-    if (check.path.startsWith(":x-pack:")) {
-      checkPart2.dependsOn check
-    } else {
-      checkPart1.dependsOn check
+    task checkPart1
+    task checkPart2
+    tasks.matching { it.name == "check" }.all { check ->
+        if (check.path.startsWith(":x-pack:")) {
+            checkPart2.dependsOn check
+        } else {
+            checkPart1.dependsOn check
+        }
     }
-  } 
 }
 
-
+subprojects {
```

Comment:
I'm not sure this "decoupling" is really all that much better. I certainly think the registering of artifacts bit is better but realitically, these things _are_ coupled, we've just moved the logic into a more obscure place.

I think the awkwardness is around how we fetch an instance of `versions`. I think we could really simplify this by moving [this](https://github.com/elastic/elasticsearch/blob/71d407f13c6d82afbe4f85a5e31c5d771295145d/build.gradle#L110) into a static method on `BwcVersions`. Something like `getInstance()`. Then we could get at this from anywhere we want without having to awkwardly get at a extra property on the root project.

Reply:
Making it easier to reference will just increase the coupling. 
I don't see why you would think they are coupled, with the registering of the artifacts testclusters don't care at all where these are coming from and weather these were released or not, and the bits that do care  moved closer together in the build script. 
Also this will offer more clarity when ruining in external plugins that can only deal with released versions. I also think it makes it easier to understand and reason about what testclsuters does as it doesn't mix in related functions like dealing with unreleased artifacts.

Reply:
My concern is that's its more difficult to reason about where test clusters is resolving these artifacts from because they are registered outside the plugin. Injection in Gradle plugins is simply awkward and there's no good solution for it. So the issue with exposing some API is that it could be called from _anywhere_. This is probably fine but I generally like to avoid adding more stuff to the root `build.gradle` which is already getting large and doesn't benefit from static compilation and type safety like stuff in `buildSrc` does.